Finish implementing ICUs in the template pipeline

mmalerba opened the pull request:
Contains a series of fixes for the last few remaining issues related to ICUs. See individual commits for more details

mmalerba committed:
refactor(compiler): Put i18n expression ops in the correct order

When we re-assign the slot dependencies for the i18nExprs, we should
move them down below the other ops that target their same slot. This
keeps the behavior consistent with TDB

mmalerba committed:
refactor(compiler): Handle i18n placeholders with spaces

I18n placeholders may contain spaces, this change updates the formatting
logic to replace them with underscores in the output.

mmalerba committed:
refactor(compiler): Add support for ICUs that share a placeholder

ICUs may share a placeholder, and in that case they need special
post-processing. This change adds logic to cover this possibility. In
particular, we set the param to a special placeholder value and then
pass an array containing the sub-message variables as a post-processing
param.

mmalerba committed:
refactor(compiler): Remove invalid assertion about i18n params

We previously had an assertion that every placeholder in the i18n AST
had a corresponding param in the output. However, there are some cases
such as interpolations nested inside ICUs where this assertion is not
true. This change simply removes the asserion.

mmalerba committed:
refactor(compiler): Fix ingestion of nested ICUs

It is possible for ICUs to be nested inside other ICUs. This change
adjusts our ingestion logic to create extra interpolation ops for the
nested ICUs during ingestion.

mmalerba committed:
test(compiler): Enable newly passing tests

Enables additional tests that pass after implementing the ICU logic.

dylhunn commented on packages/compiler/src/template/pipeline/src/phases/assign_i18n_slot_dependencies.ts:
```diff
@@ -35,11 +35,44 @@ export function assignI18nSlotDependencies(job: CompilationJob) {
+    let previousTarget: ir.XrefId|null = null;
+    let currentTarget: ir.XrefId|null = null;
     for (const op of unit.update) {
+      currentTarget = ir.hasDependsOnSlotContextTrait(op) ? op.target : null;
```
Perhaps we should add a TODO here that this ordering constraint can be relaxed after we no longer care about the TDB order?

dylhunn commented on packages/compiler/src/template/pipeline/src/phases/i18n_const_collection.ts:
```diff
@@ -76,13 +84,17 @@ function collectMessage(
+    subMessages.push(subMessageVar);
+    subMessagePlaceholders.set(subMessage.messagePlaceholder!, subMessages);
   }
+  addSubMessageParams(messageOp, subMessagePlaceholders);
```
Mind adding the commit message info as comments here as well? I worry about figuring out this logic down the road

dylhunn commented on packages/compiler/src/template/pipeline/src/ingest.ts:
```diff
@@ -492,12 +492,7 @@ function ingestIcu(unit: ViewCompilationUnit, icu: t.Icu) {
-      throw Error('ICU should have a text binding');
-    }
-    ingestBoundText(unit, icu.vars[expressionPlaceholder], [expressionPlaceholder]);
-    for (const [placeholder, text] of Object.entries(icu.placeholders)) {
```
ICU-ception 😱

mmalerba committed:
docs(compiler): Improve code comments

Adds better comments describing the case where we have multiple ICU
sub-messages that share the same plaeholder.

dylhunn commented:
This PR was merged into the repository by commit acd6100960a17e5425f3895883bc91f1499ecb28.
